docs/debugging-reconfigurator.adoc

```diff
@@ -0,0 +1,1466 @@
+
+Reducing the duration upgrade is an area of ongoing engineering focus.
+
+=== Impact on running instances
+
+As of this writing, the system does not do anything to minimize the impact of upgrade on running instances.  Instances are affected primarily in two ways: when the sled hosting the instance itself is rebooted for upgrade and when the Crucible downstairs instances backing one of the instance's disks are offline, either because of a zone reboot or sled reboot.
+
+When the sled hosting the instance itself reboots, the instance will not be running between when the sled is rebooted and when the instance starts again.  The instance is not restarted until after the sled it was on has finished booting.footnote:[Again, because the update system doesn't do anything to minimize disruption, this condition appears to the rest of the system indistinguishable from a _partition_ of the sled hosting the instance.  That's why it doesn't take action until the sled comes back.]  This can happen to an instance more than once during an upgrade, if the sled it gets restarted on itself needs to be rebooted for its upgrade.  In general, this will happen between 1 and `N-1` times for each instance, where `N` is the number of sleds in the system.  The instance will generally be offline for 8-13 minutes.footnote:[For more on this range of time, see https://github.com/oxidecomputer/omicron/issues/9094\[omicron#9094\].  The short version is that the first restart can itself fail spuriously (due to a bug) and it may take an extra 5 minutes before the system tries to restart it again.]
+
+When two of the Crucible downstairs instances backing one of the instance's disks become unavailable, the instance may see I/O delays, timeouts, or other errors until one of the instances becomes available again.  For Crucible zone updates, this is about a minute.  For sled reboots, this is closer to 10 minutes.  Keep in mind that this only happens when _multiple_ Crucible zones are affected at the same time (generally two) and the problem resolves when all but one come back (generally just one needs to come back).
+
+Reducing the duration and impact of the disruption is an area of ongoing engineering focus.
+
+=== Impact on the API, CLI, and console
+
+Broadly speaking, the API, CLI, and console should be working for the duration of the upgrade _except_ during Nexus handoff.  Handoff goes through a few phases:
+
+. First, new sagas are disallowed.  API operations that create sagas (e.g., starting/stopping instances) will fail starting at this point.  However, other operations will succeed  (e.g., listing projects and instances, creating ssh keys, etc.). Next, existing sagas must complete.
+. Once existing sagas complete, existing Nexus instances cease use of the database.  _All_ API requests will begin failing at this point.
+. Shortly after that, the new Nexus instances will take over and resume normal operation.
```

Comment:
Might be worth mentioning schema migration here too. I'm vaguely remembering one release where a migration step was quite slow, which would've widened this window a fair bit.